v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/convertors/ChangeEventSessionConvertor.java

```diff
@@ -90,8 +103,10 @@ public JsonNode transformChangeEventViaSessionFile(JsonNode changeEvent) {
     // Remove columns present in change event that were dropped in Spanner.
     changeEvent = removeDroppedColumns(changeEvent, tableId);
 
-    // Add shard id to change event.
-    changeEvent = populateShardId(changeEvent, tableId);
+    if (transformationContext != null) {
+      // Add shard id to change event.
+      changeEvent = populateShardId(changeEvent, tableId);
```

Comment:
where is the shardId being fetched from ?

Reply:
ShardId is fetched from the transformation context in the forward migration template. However, since there is no transformationContext (yet) in bulk template, this code wont be executed for bulk. The if statement should handle it.

Reply:
Test cases already exist for this in master.